src/passes/CastRefining.cpp

```diff
@@ -0,0 +1,292 @@
+  }
+
+  void computeAbstractTypes(const SubTypes& subTypes) {
+    // Abstract types are those with no news, i.e., the complement of
+    // |createdTypes|. As mentioned above, we can only optimize this case if
+    // traps never happen.
+    Types abstractTypes;
+    for (auto type : subTypes.types) {
+      if (createdTypes.count(type) == 0) {
+        abstractTypes.insert(type);
+      }
+    }
+
+    // We found abstract types. Next, find which of them are refinable. We
+    // need an abstract type to have a single subtype, to which we will switch
+    // all of their casts.
+    //
+    // Do this depth-first, so that we visit subtypes first. That will handle
+    // chains where we want to refine a type A to a subtype of a subtype of
+    // it.
+    for (auto type : subTypes.getSubTypesFirstSort()) {
+      if (!abstractTypes.count(type)) {
+        continue;
+      }
+
+      std::optional<HeapType> refinedType;
+      auto& typeSubTypes = subTypes.getStrictSubTypes(type);
+      if (typeSubTypes.size() == 1) {
+        // There is only a single possibility, so we can definitely use that
+        /// one.
+        refinedType = typeSubTypes[0];
+      } else if (!typeSubTypes.empty()) {
+        // There are multiple possibilities. However, perhaps only one of them
+        // is relevant, if nothing is ever created of the others or their
+        // subtypes.
+        for (auto subType : typeSubTypes) {
+          if (createdTypesOrSubTypes.count(subType)) {
+            if (!refinedType) {
+              // This is the first relevant thing, and hopefully will remain
+              // the only one.
+              refinedType = subType;
+            } else {
+              // We've seen more than one as relevant, so we have failed to
+              // find a singleton.
+              refinedType = std::nullopt;
+              break;
+            }
+          }
+        }
+      }
+      if (refinedType) {
+        // Propagate anything from the child, to handle chains.
+        auto iter = refinableTypes.find(*refinedType);
+        if (iter != refinableTypes.end()) {
+          *refinedType = iter->second;
+        }
+
+        refinableTypes[type] = *refinedType;
+      }
+    }
+  }
+
+  void optimize(Module* module, const SubTypes& subTypes) {
+    // To optimize we rewrite types. That is, if we want to optimize all casts
+    // of $A to instead cast to the refined type $B, we can do that by simply
+    // replacing all appearances of $A with $B. That is possible here since we
+    // only optimize when we know $A is never created, and we are removing all
+    // casts to it, which means no other references to it are needed - so we can
+    // just rewrite all references to $A to point to $B. Doing such a rewrite
+    // will also remove the unneeded type from the type section, which is nice
+    // for code size.
+    //
+    // Note that this could in theory limit optimizability, but it does not. It
```

Comment:
Maybe rather than saying it limits optimizability "in theory" but not actually, which is somewhat confusing, it would be better to say that it "may allow a later TypeMerging to merge more types, which could inhibit more optimizations after that pass."

Reply:
The possible issue is even without TypeMerging, though. TypeMerging is risky (can limit opts), and this paragraph is trying to say that **this** optimization may appear to have the same risk - but in fact it does not.

Reply:
Maybe something like "Even though this pass removes types, it does not on its own inhibit further optimizations" would be clearer, then.

Reply:
Sounds good, rewritten with that base.